java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaClient.java

```diff
@@ -313,66 +325,133 @@ public void delete(final String rawPath) throws IOException {
+                                throw new UncheckedIOException(e);
+                            }
+                        /* This exception can be thrown if the parallelism value
+                         * isn't tuned for the findForkJoinPool in relation to
+                         * the amount of bandwidth available. Essentially, the
+                         * processing thread is waiting too long for a new
+                         * connection from the pool. If this is thrown too often,
+                         * the maximum number of connections can be increased,
+                         * the ConnectionRequestTimeout can be increased, or
+                         * the fork join pool parallelism value can be
+                         * decreased.
+                         * Below we cope with this problem, by skipping the
+                         * deletion of the object and letting it
+                         * get deleted later in the loop when there is less
+                         * contention on the connection pool.
+                         */
+                        } catch (ConnectionPoolTimeoutException e) {
+                            if (LOG.isDebugEnabled()) {
+                                String msg = e.getMessage()
+                                        + " for deleting object {}";
+                                LOG.debug(msg, obj.getPath());
+                                return obj;
+                            }
+                        } catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+
+                        obj.getHttpHeaders().put("deleted", true);
+
+                        return obj;
+                    })
+                    /* We then sort the directories (and remaining files) with
+                     * the deepest paths in the filesystem hierarchy first, so
+                     * that we can delete subdirectories and files before
+                     * the parent directories.*/
+                    .sorted(MantaObjectDepthComparator.INSTANCE);
+
+            /* We go through every remaining directory and file attempt to
+             * delete it even though that operation may not be immediately
+             * successful. */
+            toDelete.forEach(obj -> {
+                for (int i = 0; i < config.getRetries(); i++) {
+                    try {
+                        /* Don't bother deleting the file if it was marked as
+                         * deleted from the map step. */
+                        if (obj.getHttpHeaders().containsKey("deleted")) {
+                            break;
+                        }
+
+                        /* If a file snuck in, we will delete it here. Typically
+                         * this should be an empty directory. */
+                        delete(obj.getPath());
+
+                        if (LOG.isTraceEnabled()) {
```

Comment:
nit: I don't think ` obj.getPath()` is so expensive that it merits a double level guard.

Reply:
You are write it is ugly. I prefer the guard clause because `obj.getPath()` might become more expensive and then someone would have to remember to add guard clauses to all usages.
Honestly, that is a weakly held position that I hold with no passion.